src/Backend/Core/Installer/AbstractModuleInstaller.php

```diff
@@ -0,0 +1,1056 @@
+    {
+        $this->defaultExtras[] = ['id' => $extraId, 'position' => $position];
+    }
+
+    /**
+     * Inserts a new module.
+     * The getModule method becomes available after using addModule and returns $module parameter.
+     *
+     * @param string $module The name of the module.
+     */
+    protected function addModule(string $module): void
+    {
+        $this->module = (string) $module;
+
+        // module does not yet exists
+        if (!(bool) $this->getDatabase()->getVar('SELECT 1 FROM modules WHERE name = ? LIMIT 1', $this->module)) {
+            // build item
+            $item = [
+                'name' => $this->module,
+                'installed_on' => gmdate('Y-m-d H:i:s'),
+            ];
+
+            // insert module
+            $this->getDatabase()->insert('modules', $item);
+
+            return;
+        }
+
+        // activate and update description
+        $this->getDatabase()->update('modules', ['installed_on' => gmdate('Y-m-d H:i:s')], 'name = ?', $this->module);
+    }
+
+    /**
+     * Add a search index
+     *
+     * @param string $module The module wherein will be searched.
+     * @param int $otherId The id of the record.
+     * @param array $fields A key/value pair of fields to index.
+     * @param string $language The frontend language for this entry.
+     */
+    protected function addSearchIndex(string $module, int $otherId, array $fields, string $language): void
+    {
+        // get database
+        $database = $this->getDatabase();
+
+        // validate cache
+        if (empty(self::$modules)) {
+            // get all modules
+            self::$modules = (array) $database->getColumn('SELECT m.name FROM modules AS m');
+        }
+
+        // module exists?
+        if (!in_array('Search', self::$modules)) {
+            return;
+        }
+
+        // no fields?
+        if (empty($fields)) {
+            return;
+        }
+
+        // insert search index
+        foreach ($fields as $field => $value) {
+            // reformat value
+            $value = strip_tags((string) $value);
+
+            // insert in database
+            $database->execute(
+                'INSERT INTO search_index (module, other_id, language, field, value, active)
+                 VALUES (?, ?, ?, ?, ?, ?)
+                 ON DUPLICATE KEY UPDATE value = ?, active = ?',
+                [(string) $module, (int) $otherId, (string) $language, (string) $field, $value, true, $value, true]
+            );
+        }
+
+        // invalidate the cache for search
+        $finder = new Finder();
+        $filesystem = new Filesystem();
+        foreach ($finder->files()->in(FRONTEND_CACHE_PATH . '/Search/') as $file) {
+            /** @var $file \SplFileInfo */
+            $filesystem->remove($file->getRealPath());
+        }
+    }
+
+    /**
+     * Method that will be overridden by the specific installers
+     */
+    protected function execute(): void
```

Comment:
I would place this implementation in the ModuleInstaller class and make it abstract here